File: experiments-local/experiment-17/runs/agent=omp_language=typescript_model=openrouter/anthropic/claude-opus-4.8_tooling=none/rep2/src/loader.ts
/**
 * Context
 * -------
 * Loads the six Kaggle CSV files in `data/kaggle/` into the unified `Match`
 * and `Player` models and exposes them through a `Dataset` object.
 *
 * Responsibilities:
 *   - Map each file's columns onto the canonical shape.
 *   - Assign a canonical `Competition` label per row.
 *   - Deduplicate matches that appear in more than one source (Série A and
 *     Copa do Brasil are present both in the dedicated files and in
 *     BR-Football-Dataset.csv). The richer "extended statistics" source merges
 *     its corner/shot/half-time data into the already-loaded match instead of
 *     creating a duplicate.
 *
 * Loading is synchronous and one-shot; with ~33k matches + 18k players it
 * comfortably fits the spec's <2s simple / <5s aggregate latency budget.
 */

import { readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

import { parseCsv } from "./csv.js";
import { parseDate, parseTeam } from "./normalize.js";
import type { Competition, Match, MatchStats, Player, SourceFile } from "./models.js";

const HERE = dirname(fileURLToPath(import.meta.url));
/** Default data directory: `<repo>/data/kaggle` regardless of cwd. */
export const DEFAULT_DATA_DIR = join(HERE, "..", "data", "kaggle");

export interface Dataset {
  readonly matches: readonly Match[];
  readonly players: readonly Player[];
}

/** Numeric coercion that tolerates blanks, quotes, and trailing ".0" floats. */
function num(value: string | undefined): number | undefined {
  if (value === undefined) return undefined;
  const t = value.trim().replace(/^"|"$/g, "");
  if (t === "") return undefined;
  const n = Number(t);
  return Number.isFinite(n) ? n : undefined;
}

function makeMatch(args: {
  source: SourceFile;
  competition: Competition;
  season: number;
  date?: string;
  round?: string;
  stage?: string;
  stadium?: string;
  home: string;
  away: string;
  homeGoals: number;
  awayGoals: number;
  stats?: MatchStats;
}): Match {
  return {
    source: args.source,
    competition: args.competition,
    season: args.season,
    date: parseDate(args.date),
    round: args.round?.trim() || undefined,
    stage: args.stage?.trim() || undefined,
    stadium: args.stadium?.trim() || undefined,
    homeTeamRaw: args.home,
    awayTeamRaw: args.away,
    home: parseTeam(args.home),
    away: parseTeam(args.away),
    homeGoals: args.homeGoals,
    awayGoals: args.awayGoals,
    stats: args.stats,
  };
}

const BR_FOOTBALL_COMPETITION: Record<string, Competition> = {
  "serie a": "Brasileirão Série A",
  "serie b": "Brasileirão Série B",
  "serie c": "Brasileirão Série C",
  "copa do brasil": "Copa do Brasil",
};

function loadBrasileirao(dir: string): Match[] {
  const rows = parseCsv(readFileSync(join(dir, "Brasileirao_Matches.csv"), "utf8"));
  const out: Match[] = [];
  for (const r of rows) {
    const hg = num(r["home_goal"]);
    const ag = num(r["away_goal"]);
    const season = num(r["season"]);
    if (hg === undefined || ag === undefined || season === undefined) continue;
    out.push(
      makeMatch({
        source: "Brasileirao_Matches.csv",
        competition: "Brasileirão Série A",
        season,
        date: r["datetime"],
        round: r["round"],
        home: r["home_team"] ?? "",
        away: r["away_team"] ?? "",
        homeGoals: hg,
        awayGoals: ag,
      }),
    );
  }
  return out;
}

function loadCup(dir: string): Match[] {
  const rows = parseCsv(readFileSync(join(dir, "Brazilian_Cup_Matches.csv"), "utf8"));
  const out: Match[] = [];
  for (const r of rows) {
    const hg = num(r["home_goal"]);
    const ag = num(r["away_goal"]);
    const season = num(r["season"]);
    if (hg === undefined || ag === undefined || season === undefined) continue;
    out.push(
      makeMatch({
        source: "Brazilian_Cup_Matches.csv",
        competition: "Copa do Brasil",
        season,
        date: r["datetime"],
        round: r["round"],
        home: r["home_team"] ?? "",
        away: r["away_team"] ?? "",
        homeGoals: hg,
        awayGoals: ag,
      }),
    );
  }
  return out;
}

function loadLibertadores(dir: string): Match[] {
  const rows = parseCsv(readFileSync(join(dir, "Libertadores_Matches.csv"), "utf8"));
  const out: Match[] = [];
  for (const r of rows) {
    const hg = num(r["home_goal"]);
    const ag = num(r["away_goal"]);
    const season = num(r["season"]);
    if (hg === undefined || ag === undefined || season === undefined) continue;
    out.push(
      makeMatch({
        source: "Libertadores_Matches.csv",
        competition: "Copa Libertadores",
        season,
        date: r["datetime"],
        stage: r["stage"],
        home: r["home_team"] ?? "",
        away: r["away_team"] ?? "",
        homeGoals: hg,
        awayGoals: ag,
      }),
    );
  }
  return out;
}

function loadHistorical(dir: string): Match[] {
  const rows = parseCsv(readFileSync(join(dir, "novo_campeonato_brasileiro.csv"), "utf8"));
  const out: Match[] = [];
  for (const r of rows) {
    const hg = num(r["Gols_mandante"]);
    const ag = num(r["Gols_visitante"]);
    const season = num(r["Ano"]);
    if (hg === undefined || ag === undefined || season === undefined) continue;
    out.push(
      makeMatch({
        source: "novo_campeonato_brasileiro.csv",
        competition: "Brasileirão Série A",
        season,
        date: r["Data"],
        round: r["Rodada"],
        stadium: r["Arena"],
        home: r["Equipe_mandante"] ?? "",
        away: r["Equipe_visitante"] ?? "",
        homeGoals: hg,
        awayGoals: ag,
      }),
    );
  }
  return out;
}

function loadExtended(dir: string): Match[] {
  const rows = parseCsv(readFileSync(join(dir, "BR-Football-Dataset.csv"), "utf8"));
  const out: Match[] = [];
  for (const r of rows) {
    const hg = num(r["home_goal"]);
    const ag = num(r["away_goal"]);
    const competition = BR_FOOTBALL_COMPETITION[(r["tournament"] ?? "").trim().toLowerCase()];
    const date = parseDate(r["date"]);
    if (hg === undefined || ag === undefined || !competition || !date) continue;
    const stats: MatchStats = {
      homeCorners: num(r["home_corner"]),
      awayCorners: num(r["away_corner"]),
      totalCorners: num(r["total_corners"]),
      homeShots: num(r["home_shots"]),
      awayShots: num(r["away_shots"]),
      homeAttacks: num(r["home_attack"]),
      awayAttacks: num(r["away_attack"]),
    };
    out.push(
      makeMatch({
        source: "BR-Football-Dataset.csv",
        competition,
        season: date.year,
        date: r["date"],
        home: r["home"] ?? "",
        away: r["away"] ?? "",
        homeGoals: hg,
        awayGoals: ag,
        stats,
      }),
    );
  }
  return out;
}

/**
 * Per-source preference rank. For any given (competition, season) we keep the
 * matches from the single highest-ranked source that covers it, instead of
 * fuzzy-deduplicating near-identical rows whose team spellings diverge between
 * sources ("Atletico-PR" vs "Athletico-PR"). This guarantees each real game is
 * counted exactly once — critical for standings and win-rate math.
 *
 *   - Série A 2012–2022 → Brasileirao_Matches.csv (rank 4)
 *   - Série A 2003–2011 → novo_campeonato_brasileiro.csv (rank 3)
 *   - Série A 2023, Série B/C → BR-Football-Dataset.csv (rank 1, sole source)
 *   - Copa do Brasil → Brazilian_Cup_Matches.csv (rank 4)
 *   - Copa Libertadores → Libertadores_Matches.csv (rank 4)
 */
const SOURCE_RANK: Record<SourceFile, number> = {
  "Brasileirao_Matches.csv": 4,
  "Brazilian_Cup_Matches.csv": 4,
  "Libertadores_Matches.csv": 4,
  "novo_campeonato_brasileiro.csv": 3,
  "BR-Football-Dataset.csv": 1,
};

/** Stats-enrichment key: competition, season, folded teams, score. */
function statsKey(m: Match): string {
  return [m.competition, m.season, m.home.baseKey, m.away.baseKey, m.homeGoals, m.awayGoals].join("|");
}

/**
 * Select the canonical match list:
 *   1. For each (competition, season) keep only rows from the top-ranked source.
 *   2. Enrich surviving rows with extended stats from BR-Football-Dataset.csv
 *      (corners/shots/attacks), matched on {competition, season, teams, score}.
 */
function selectCanonical(all: Match[], extended: Match[]): Match[] {
  const bestRank = new Map<string, number>();
  for (const m of all) {
    const g = `${m.competition}|${m.season}`;
    const rank = SOURCE_RANK[m.source];
    if (rank > (bestRank.get(g) ?? 0)) bestRank.set(g, rank);
  }

  const canonical = all.filter((m) => SOURCE_RANK[m.source] === bestRank.get(`${m.competition}|${m.season}`));

  // Index every extended row's stats so non-canonical sources still contribute
  // their statistics to the canonical match for the same game.
  const stats = new Map<string, MatchStats>();
  for (const m of extended) {
    if (m.stats) stats.set(statsKey(m), m.stats);
  }
  for (const m of canonical) {
    if (!m.stats) {
      const found = stats.get(statsKey(m));
      if (found) m.stats = found;
    }
  }
  return canonical;
}

function loadPlayers(dir: string): Player[] {
  const rows = parseCsv(readFileSync(join(dir, "fifa_data.csv"), "utf8"));
  const out: Player[] = [];
  for (const r of rows) {
    const id = num(r["ID"]);
    const name = (r["Name"] ?? "").trim();
    if (id === undefined || !name) continue;
    out.push({
      id,
      name,
      age: num(r["Age"]),
      nationality: (r["Nationality"] ?? "").trim(),
      overall: num(r["Overall"]),
      potential: num(r["Potential"]),
      club: (r["Club"] ?? "").trim(),
      position: (r["Position"] ?? "").trim(),
      jerseyNumber: num(r["Jersey Number"]),
      height: (r["Height"] ?? "").trim() || undefined,
      weight: (r["Weight"] ?? "").trim() || undefined,
      preferredFoot: (r["Preferred Foot"] ?? "").trim() || undefined,
    });
  }
  return out;
}

/** Load every dataset from `dir` (defaults to the bundled `data/kaggle`). */
export function loadDataset(dir: string = DEFAULT_DATA_DIR): Dataset {
  const extended = loadExtended(dir);
  const all = [
    ...loadBrasileirao(dir),
    ...loadCup(dir),
    ...loadLibertadores(dir),
    ...loadHistorical(dir),
    ...extended,
  ];
  const matches = selectCanonical(all, extended);
  const players = loadPlayers(dir);
  return { matches, players };
}

File: experiments-local/experiment-17/runs/agent=omp_language=python_model=openrouter/anthropic/claude-opus-4.8_tooling=none/rep2/brazilian_soccer_mcp/loader.py
"""
Context
=======
Module: brazilian_soccer_mcp.loader
Purpose: Read the six provided Kaggle CSVs and project them into two normalized
         in-memory pandas tables: a unified ``matches`` frame and a ``players``
         frame. All team names are canonicalized via :mod:`.normalize` so that
         cross-source queries (head-to-head, standings) are consistent.

Unified match schema (one row per match)
----------------------------------------
    date         datetime64  -- kickoff date (time dropped; multiple input formats)
    season       int         -- competition year
    competition  str         -- canonical competition name (see _COMP_*)
    home_raw     str         -- original home name (for display)
    away_raw     str
    home_key     str         -- normalized lookup key (suffix-stripped, deaccented)
    away_key     str
    home_state   str|None    -- state/country token when present
    away_state   str|None
    home_goal    int
    away_goal    int
    stage        str|None    -- round/stage label for display
    home_shots   float       -- extended stats (BR-Football only; else NaN)
    away_shots   float
    home_corner  float
    away_corner  float
    source       str         -- originating CSV basename

Source partitioning (instead of cross-source dedup)
----------------------------------------------------
Three sources overlap on Brasileirão Série A (Brasileirao_Matches 2012-2022,
novo_campeonato 2003-2019, BR-Football "Serie A"), but they spell the same club
irreconcilably differently ("Vasco" / "Vasco da Gama RJ", "Athletico" /
"Atlético Paranaense"), so the same fixture gets different normalized keys and
row-level dedup cannot collapse it. Instead each ``(competition, season)`` slice
is sourced from exactly ONE CSV — the highest-priority one present (see
``_SOURCE_PRIORITY``) — guaranteeing clean, non-double-counted round-robin
standings and head-to-head records. A final intra-source key dedup removes any
accidental repeats within a single file.
"""

from __future__ import annotations

from pathlib import Path

import pandas as pd

from .normalize import base_key, split_suffix, team_key

# Canonical competition labels.
_COMP_SERIE_A = "Brasileirão Série A"
_COMP_SERIE_B = "Brasileirão Série B"
_COMP_SERIE_C = "Brasileirão Série C"
_COMP_COPA_BR = "Copa do Brasil"
_COMP_LIBERTADORES = "Copa Libertadores"

_BR_FOOTBALL_TOURNAMENT = {
    "Serie A": _COMP_SERIE_A,
    "Serie B": _COMP_SERIE_B,
    "Serie C": _COMP_SERIE_C,
    "Copa do Brasil": _COMP_COPA_BR,
}

_MATCH_COLUMNS = [
    "date", "season", "competition",
    "home_raw", "away_raw", "home_key", "away_key", "home_id", "away_id",
    "home_state", "away_state", "home_goal", "away_goal",
    "stage", "home_shots", "away_shots", "home_corner", "away_corner",
    "source",
]


def _key_series(raw: pd.Series) -> pd.Series:
    """Vectorized ``base_key`` (state-stripped) over a name column, computing each
    distinct raw string only once."""
    uniques = raw.dropna().unique()
    table = {u: base_key(u) for u in uniques}
    return raw.map(table)


def _id_series(raw: pd.Series) -> pd.Series:
    """Vectorized ``team_key`` (state-aware identity) over a name column."""
    uniques = raw.dropna().unique()
    table = {u: team_key(u) for u in uniques}
    return raw.map(table)


def _state_series(raw: pd.Series) -> pd.Series:
    uniques = raw.dropna().unique()
    table = {u: split_suffix(u)[1] for u in uniques}
    return raw.map(table)


def _finalize(df: pd.DataFrame, source: str) -> pd.DataFrame:
    """Coerce goals to int, drop rows lacking a result, add normalized keys and
    any missing optional columns, and stamp the source."""
    df = df.copy()
    df["home_goal"] = pd.to_numeric(df["home_goal"], errors="coerce")
    df["away_goal"] = pd.to_numeric(df["away_goal"], errors="coerce")
    df = df.dropna(subset=["home_goal", "away_goal", "home_raw", "away_raw"])
    df["home_goal"] = df["home_goal"].astype(int)
    df["away_goal"] = df["away_goal"].astype(int)
    df["home_key"] = _key_series(df["home_raw"])
    df["away_key"] = _key_series(df["away_raw"])
    df["home_id"] = _id_series(df["home_raw"])
    df["away_id"] = _id_series(df["away_raw"])
    df["home_state"] = _state_series(df["home_raw"])
    df["away_state"] = _state_series(df["away_raw"])
    df["source"] = source
    for col in ("stage", "home_shots", "away_shots", "home_corner", "away_corner"):
        if col not in df.columns:
            df[col] = pd.NA
    df["season"] = pd.to_numeric(df["season"], errors="coerce").astype("Int64")
    df = df.dropna(subset=["season"])
    df["season"] = df["season"].astype(int)
    return df[_MATCH_COLUMNS]


def _load_brasileirao(path: Path) -> pd.DataFrame:
    df = pd.read_csv(path)
    out = pd.DataFrame({
        "date": pd.to_datetime(df["datetime"], errors="coerce"),
        "season": df["season"],
        "competition": _COMP_SERIE_A,
        "home_raw": df["home_team"],
        "away_raw": df["away_team"],
        "home_goal": df["home_goal"],
        "away_goal": df["away_goal"],
        "stage": "Round " + df["round"].astype(str),
    })
    return _finalize(out, "Brasileirao_Matches.csv")


def _load_cup(path: Path) -> pd.DataFrame:
    df = pd.read_csv(path)
    out = pd.DataFrame({
        "date": pd.to_datetime(df["datetime"], errors="coerce"),
        "season": df["season"],
        "competition": _COMP_COPA_BR,
        "home_raw": df["home_team"],
        "away_raw": df["away_team"],
        "home_goal": df["home_goal"],
        "away_goal": df["away_goal"],
        "stage": df["round"].astype(str),
    })
    return _finalize(out, "Brazilian_Cup_Matches.csv")


def _load_libertadores(path: Path) -> pd.DataFrame:
    df = pd.read_csv(path)
    out = pd.DataFrame({
        "date": pd.to_datetime(df["datetime"], errors="coerce"),
        "season": df["season"],
        "competition": _COMP_LIBERTADORES,
        "home_raw": df["home_team"],
        "away_raw": df["away_team"],
        "home_goal": df["home_goal"],
        "away_goal": df["away_goal"],
        "stage": df["stage"],
    })
    return _finalize(out, "Libertadores_Matches.csv")


def _load_novo(path: Path) -> pd.DataFrame:
    df = pd.read_csv(path)
    out = pd.DataFrame({
        "date": pd.to_datetime(df["Data"], format="%d/%m/%Y", errors="coerce"),
        "season": df["Ano"],
        "competition": _COMP_SERIE_A,
        "home_raw": df["Equipe_mandante"],
        "away_raw": df["Equipe_visitante"],
        "home_goal": df["Gols_mandante"],
        "away_goal": df["Gols_visitante"],
        "stage": "Round " + df["Rodada"].astype(str),
    })
    return _finalize(out, "novo_campeonato_brasileiro.csv")


def _load_br_football(path: Path) -> pd.DataFrame:
    df = pd.read_csv(path)
    date = pd.to_datetime(df["date"], errors="coerce")
    out = pd.DataFrame({
        "date": date,
        "season": date.dt.year,
        "competition": df["tournament"].map(_BR_FOOTBALL_TOURNAMENT),
        "home_raw": df["home"],
        "away_raw": df["away"],
        "home_goal": df["home_goal"],
        "away_goal": df["away_goal"],
        "stage": pd.NA,
        "home_shots": df.get("home_shots"),
        "away_shots": df.get("away_shots"),
        "home_corner": df.get("home_corner"),
        "away_corner": df.get("away_corner"),
    })
    out = out.dropna(subset=["competition"])
    return _finalize(out, "BR-Football-Dataset.csv")


# Per-source priority. For any (competition, season) slice covered by more than
# one CSV we keep exactly ONE source — the lowest-priority number present —
# because the three Série A sources spell teams irreconcilably differently
# ("Vasco" vs "Vasco da Gama RJ", "Athletico" vs "Atlético Paranaense"), so
# key-based row dedup across sources cannot collapse the same fixture. Choosing
# a single source per slice guarantees clean round-robin standings/H2H.
_SOURCE_PRIORITY = {
    "Brasileirao_Matches.csv": 1,       # Série A 2012-2022, round-annotated
    "Brazilian_Cup_Matches.csv": 1,     # Copa do Brasil 2012-2021
    "Libertadores_Matches.csv": 1,      # Libertadores 2013-2022
    "novo_campeonato_brasileiro.csv": 2,  # Série A 2003-2019 (historical fill)
    "BR-Football-Dataset.csv": 3,       # fallback; sole source for Série B/C & 2023
}


def load_matches(data_dir: Path) -> pd.DataFrame:
    """Load and unify all match CSVs into a single non-overlapping frame.

    Each (competition, season) is sourced from exactly one CSV (the highest
    priority present), so no cross-source duplicate fixtures can occur. A final
    intra-source dedup on normalized keys drops any accidental repeats within a
    single file.
    """
    frames = [
        _load_brasileirao(data_dir / "Brasileirao_Matches.csv"),
        _load_novo(data_dir / "novo_campeonato_brasileiro.csv"),
        _load_cup(data_dir / "Brazilian_Cup_Matches.csv"),
        _load_libertadores(data_dir / "Libertadores_Matches.csv"),
        _load_br_football(data_dir / "BR-Football-Dataset.csv"),
    ]
    matches = pd.concat(frames, ignore_index=True)
    matches["_prio"] = matches["source"].map(_SOURCE_PRIORITY)
    best = matches.groupby(["competition", "season"])["_prio"].transform("min")
    matches = matches[matches["_prio"] == best]
    matches = matches.drop_duplicates(
        subset=["competition", "season", "home_id", "away_id"],
        keep="first",
    ).drop(columns="_prio").reset_index(drop=True)
    return matches


def load_players(data_dir: Path) -> pd.DataFrame:
    """Load the FIFA player database, retaining display + filter columns and a
    normalized club key for cross-file club matching."""
    df = pd.read_csv(data_dir / "fifa_data.csv")
    cols = ["Name", "Age", "Nationality", "Overall", "Potential",
            "Club", "Position", "Jersey Number", "Height", "Weight"]
    keep = [c for c in cols if c in df.columns]
    players = df[keep].copy()
    players["club_key"] = _key_series(players["Club"].fillna(""))
    players["name_lower"] = players["Name"].fillna("").str.lower()
    return players
